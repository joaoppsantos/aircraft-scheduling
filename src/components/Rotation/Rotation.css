.rotation::-webkit-scrollbar {
    display: none;
}

.rotation {
    height: calc(100vh - 15rem);
    width: calc(25vh);
    padding-top: 10px;
    background-color: lavender;
    margin-right: 20px;
    overflow-y: scroll;

    scrollbar-width: none; 
}

@media screen and (max-width: 670px) {
	.rotation {
        width: 90%;
        margin: 0 auto;
	}
}
