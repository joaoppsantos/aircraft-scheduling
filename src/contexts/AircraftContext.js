import { createContext, useEffect, useState } from 'react';
import request from '../request';

export const AircraftContext = createContext({});

export function AircraftProvider({ children }) {
	const [aircrafts, setAircrafts] = useState([]);
	const [isAircraftLoading, setIsAircraftLoading] = useState(true);
	const [currentAircraft, setCurrentAircraft] = useState(undefined);
	const [currentPercentage, setCurrentPercentage] = useState(0);

	const setPercentage = percentage => {
		setCurrentPercentage(percentage);
	};
	const selectAircraft = aircraft => {
		setCurrentAircraft(aircraft);
	};

	useEffect(() => {
		request.get(`/aircrafts?limit=5&offset=0`).then((response) => {
			setAircrafts(response.data.data);
			setIsAircraftLoading(false);
		});
	}, []);

	return (
		<AircraftContext.Provider value={{
			aircrafts,
			isAircraftLoading,
			currentAircraft,
			selectAircraft,
			currentPercentage,
			setPercentage
		}}>
			{children}
		</AircraftContext.Provider>
	)
}
