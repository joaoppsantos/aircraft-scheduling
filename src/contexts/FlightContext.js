import request from '../request';
import { createContext, useEffect, useState } from 'react';

export function FlightProvider({ children }) {
	const [flights, setFlights] = useState([]);
	const [activeFlights, setActiveFlights] = useState([]);
	const [offset, setOffset] = useState(-1);
	const [hasFlightsRemaining, setHasFlightsRemaining] = useState(true);
	const paginationLimit = 25;

	const toggleActiveFlight = (flight) => {
		let isFlightOnList = false;

		activeFlights.forEach((activeInfo) => {
			if (activeInfo.id === flight.id) {
				isFlightOnList = true;
			}
		});

		isFlightOnList ? 
			setActiveFlights(activeFlights.filter(activeInfo => activeInfo.id !== flight.id))
			: setActiveFlights([...activeFlights, flight]);
	}

	useEffect(() => {
		if (!hasFlightsRemaining) {
			return;
		}

		async function fetchData() {
			try {
				const newOffset = offset + 1;
				const response = await request.get(`/flights?limit=${paginationLimit}&offset=${newOffset}`);

				setOffset(newOffset);
				setFlights([...flights, ...response?.data?.data]);

				if (paginationLimit * (newOffset + 1) > response?.data?.pagination?.total) {
					setHasFlightsRemaining(false);
				}
			} catch (error) {
				console.log('Failed to load !');
			};
		}

		fetchData();
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [flights, hasFlightsRemaining]);

	return (
		<FlightContext.Provider value={{
			flights,
			toggleActiveFlight,
			activeFlights
		}}>
			{ children }
		</FlightContext.Provider>
	)
}

export const FlightContext = createContext({});