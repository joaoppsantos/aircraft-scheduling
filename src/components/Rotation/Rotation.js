import './Rotation.css';
import { useContext } from 'react';
import { FlightContext } from '../../contexts/FlightContext';
import { FlightCard } from '../FlightCard/';

const Rotation = () => {
    const {
        activeFlights,
      } = useContext(FlightContext);

    return (
        <div className="rotation">
            Rotation
			{activeFlights.map((flight, index) => {
                return <FlightCard key={index} {...flight} />
            })}
        </div>
    );
}

export default Rotation;
