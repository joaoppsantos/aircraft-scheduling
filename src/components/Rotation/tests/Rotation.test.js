import * as React from 'react';
import { render } from '@testing-library/react';
import Rotation from '../Rotation';
import { FlightContext } from '../../../contexts/FlightContext';
import { flightMock } from '../../../tests/mocks/mocks';

describe('<Rotation />', () => {
    it('should render correctly', () => {
        const mockFlightContext = {
            activeFlights: [
                flightMock,
            ],
        };
        
        const { container } = render(
            <FlightContext.Provider value={mockFlightContext}>
                <Rotation />
            </FlightContext.Provider>
        );

        expect(container).toMatchSnapshot();
    });
});
