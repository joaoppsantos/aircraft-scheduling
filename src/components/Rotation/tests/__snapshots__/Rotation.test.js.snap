// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rotation /> should render correctly 1`] = `
<div>
  <div
    class="rotation"
  >
    Rotation
    <div>
      <div
        class="flight-card"
      >
        <span
          class="flight-id"
        >
          AS1001
        </span>
        <span>
          LFSB
           
          &gt;
           
          LFMN
        </span>
        <span>
          06:00
           
          &gt;
           
          07:15
        </span>
      </div>
    </div>
  </div>
</div>
`;
